juju/application.py

```diff
@@ -361,11 +363,57 @@ def update_allocation(self, allocation):
         if charm_url == self.data['charm-url']:
             raise JujuError('already running charm "%s"' % charm_url)
 
+        # Update charm
         await client_facade.AddCharm(
             url=charm_url,
             channel=channel
         )
 
+        # Update resources
+        response = await self.model.charmstore.entity(charm_url,
+                                                      channel=channel)
```

Comment:
This is potentially duplicating a store request a few lines up.  Given how unreliable and slow these requests can be, we should avoid them as much as possible.  Given that this one is unconditionally required, we should move it up above where the other call is and change them to both use the result of a single call.

Reply:
Though, maybe instead (or in addition) we should look in to setting up a caching system for charmstore requests by revision.  I think entities can't really be modified without increasing their revision, so we shouldn't have to worry too much about cache expiration.  We could even potentially support an on-disk cache to speed things up across multiple runs, though then we would have to worry about pruning the cache at some point.